File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int[][] directions={{1,0},{-1,0},{0,1},{0,-1}};
    int rows; int cols;

    public int numIslands(char[][] grid) {
     rows=grid.length;
     cols=grid[0].length;
     int count=0;
    boolean[][] visited=new boolean[rows][cols];
    for(int i=0;i<rows;i++){
        for(int j=0;j<cols;j++){
            if(visited[i][j]!=true && grid[i][j]=='1'){
                dfs(grid,i,j,visited);
                count++;
            }
        }
    }
    return count;
    }

    public void dfs(char[][] grid,int r,int c, boolean[][] visited){
        if(r<0 ||r>=rows ||c<0 ||c>=cols || grid[r][c]=='0' || visited[r][c]==true){return;}
                visited[r][c]=true;
        for(int[] d: directions){
            dfs(grid,r+d[0],c+d[1],visited);
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    directions=[(-1,0),(1,0),(0,1),(0,-1)]
    
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m=len(grid)
        n=len(grid[0])
        visited=[[False]*n for i in range(m)]
        count=0
        for i in range(m):
            for j in range(n):
                if(grid[i][j]=='1' and visited[i][j]==False):
                    self.dfs(i,j,grid,visited,m,n)
                    count+=1
        return count

        

    def dfs(self,r,c,grid,visited,m,n):
        if(r<0 or c<0 or r>=m or c>=n or grid[r][c]=='0' or visited[r][c]==True):
            return
        visited[r][c]=True
        for dx,dy in self.directions:
            self.dfs(r+dx,c+dy,grid,visited,m,n)
             


        